### PR TITLE
20907 - Pay outstanding balance switching from EFT payment method

### PR DESCRIPTION
**Pull request by ochiu**
*Issue #:*
https://github.com/bcgov/entity/issues/20907

*Description of changes:*
- removed hybrid properties from statement model, was causing circular dependency issues and the complexity was not working well with general queries
- add statement search filter to get all statements with an amounting
- add validation to block changing payment method away from EFT when there is an outstanding balance
- extend create account payment to support paying an outstanding payment as the existing logic depends on existing failed payments for scenarios like NSF
- add amount owing to invoices that are currently not part of a statement yet in account summary
- reworked payment_methods query and optimization
- moving statement search out of model into service

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: pay-api/src/pay_api/models/statement.py [hunk 1]
```diff
@@ -115,14 +116,49 @@ def payment_methods(self):
 
         return list(payment_methods)
 
+    @hybrid_property
```
Comment (seeker25) [inline pay-api/src/pay_api/models/statement.py:119]: Have you tested the performance scaling of this property over quite a few statements?

File: pay-api/src/pay_api/models/statement.py [hunk 2]
```diff
@@ -115,14 +116,49 @@ def payment_methods(self):
...
+        query = (query.filter(StatementInvoices.statement_id == self.id)
+                 .with_entities(func.sum(Invoice.total - Invoice.paid).label('total_owing')))
+
+        result = query.scalar()
+        return result if result is not None else 0
+
+    @staticmethod
+    def get_statement_owing_query():
+        """Get statement query used for amount owing."""
+        from .statement_invoices import StatementInvoices  # pylint: disable=import-outside-toplevel
```
Comment (seeker25) [inline pay-api/src/pay_api/models/statement.py:133]: not good

File: pay-api/src/pay_api/models/statement.py [hunk 3]
```diff
@@ -115,14 +116,49 @@ def payment_methods(self):
 
         return list(payment_methods)
 
+    @hybrid_property
+    def amount_owing(self):
+        """Return amount owing on statement."""
+        from .statement_invoices import StatementInvoices  # pylint: disable=import-outside-toplevel
```
Comment (seeker25) [inline pay-api/src/pay_api/models/statement.py:122]: not good

File: pay-api/src/pay_api/resources/v1/payment.py [hunk 4]
```diff
@@ -73,10 +73,15 @@ def post_account_payment(account_id: str):
             ).asdict(), HTTPStatus.CREATED
     else:
         is_retry_payment: bool = request.args.get('retryFailedPayment', 'false').lower() == 'true'
+        pay_outstanding_balance: bool = False
+
+        if flags.is_on('enable-eft-payment-method', default=False):
```
Comment (seeker25) [inline pay-api/src/pay_api/resources/v1/payment.py:78]: have you tried LD lately? is it working correctly in main? I haven't tried it

**Review by seeker25 — CHANGES_REQUESTED** [review]
Use service instead of models to go cross models

File: pay-api/src/pay_api/models/statement.py (hunk 2 above)
Comment (ochiu) [reply pay-api/src/pay_api/models/statement.py:133]: Needed to use this as part of the query existing in the model, I originally tried to do this in the statement service, but 

service --> model --> service.

Should I just refactor the model and pull out the statement search up to the service?

Comment (seeker25) [reply pay-api/src/pay_api/models/statement.py:133]: Yes

File: pay-api/src/pay_api/models/statement.py (hunk 1 above)
Comment (ochiu) [reply pay-api/src/pay_api/models/statement.py]: Removed this from hybrid property and into the service. The hybrid property was no longer suitable due to the circular dependency issues.

Tested this on a data load with an account over 1000 statements at once, seemed performant

File: pay-api/src/pay_api/models/statement.py (hunk 3 above)
Comment (ochiu) [reply pay-api/src/pay_api/models/statement.py]: removed hybrid property and updated search query

**Comment by ochiu** [comment]
Looking into failed test

File: pay-api/src/pay_api/services/payment.py [hunk 5]
```diff
@@ -588,6 +595,72 @@ def _populate(dao: PaymentModel):
...
         return payment
 
+    @staticmethod
+    def create_consolidated_invoices_payment(consolidated_invoices: List[InvoiceModel],
+                                             consolidated_line_items: List[PaymentLineItem],
+                                             cfs_account: CfsAccountModel,
+                                             pay_account: PaymentAccountModel,
+                                             invoice_total: Decimal):
+        """Create payment for consolidated invoices and update invoice references."""
+        # Create consolidated invoice
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment.py:605]: redundant

File: pay-api/src/pay_api/services/payment.py [hunk 6]
```diff
@@ -588,6 +595,72 @@ def _populate(dao: PaymentModel):
...
+                                 payment_system=PaymentSystem.PAYBC.value,
+                                 invoice_number=invoice_number,
+                                 invoice_amount=invoice_total,
+                                 payment_account_id=pay_account.id)
+
+        return payment, invoice_number
+
+    @classmethod
+    def _consolidate_invoices_and_pay(cls, auth_account_id: str) -> Payment:
+        # Find outstanding invoices and create a payment
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment.py:634]: should use """?

File: pay-api/src/pay_api/services/statement.py [hunk 7]
```diff
@@ -128,10 +129,156 @@ def asdict(self):
...
+                payment_account_history.payment_method,
+                payment_account_history.version,
+                payment_account_history.changed.label('end_date')
+            ).correlate(StatementModel)
+            .where(
+                payment_account_history.id == statement.payment_account_id
+            )
+        )
+
+        # Combine current and historical records and calculate the date range for when it was active
```
Comment (seeker25) [inline pay-api/src/pay_api/services/statement.py:160]: good comment

File: pay-api/src/pay_api/services/statement.py [hunk 8]
```diff
@@ -271,7 +433,12 @@ def get_summary(auth_account_id: str, statement_id: str = None):
         total_due = float(result.total_due) if result else 0
         oldest_overdue_date = get_local_formatted_date(result.oldest_overdue_date) \
             if result and result.oldest_overdue_date else None
+
+        # Unpaid invoice amount total that are not part of a statement yet
```
Comment (seeker25) [inline pay-api/src/pay_api/services/statement.py:437]: good comment, not obvious by reading the code

File: pay-api/setup.cfg [hunk 9]
```diff
@@ -94,7 +94,8 @@ lines_after_imports = 2
 
 
 [tool:pytest]
-addopts = --cov=src --cov-report html:htmlcov --cov-report xml:coverage.xml
+addopts = --verbose 
+#--cov=src --cov-report html:htmlcov --cov-report xml:coverage.xml
```
Comment (seeker25) [inline pay-api/setup.cfg:98]: might want to revert this?

File: pay-api/src/pay_api/services/payment.py [hunk 10]
```diff
@@ -588,6 +595,72 @@ def _populate(dao: PaymentModel):
...
+        """Create payment for consolidated invoices and update invoice references."""
+        # Create consolidated invoice
+        invoice_response = CFSService.create_account_invoice(
+            transaction_number=str(consolidated_invoices[-1].id) + '-C',
+            line_items=consolidated_line_items,
+            cfs_account=cfs_account)
+
+        invoice_number: str = invoice_response.get('invoice_number')
+
+        # Mark all invoice references to status CANCELLED, and create a new one for the new invoice number.
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment.py:613]: redundant?

File: pay-api/src/pay_api/services/payment_account.py [hunk 11]
```diff
@@ -378,6 +378,13 @@ def _check_and_handle_payment_method(cls, account: PaymentAccountModel, target_p
                 PaymentMethod.EFT.value not in {account.payment_method, target_payment_method}:
             return
 
+        # Don't allow payment method change from EFT if there is an outstanding balance
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment_account.py:381]: redundant?

**Review by seeker25 — APPROVED** [review]

**Pull request merged**